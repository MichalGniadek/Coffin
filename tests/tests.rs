use coffin2::{debug_print, lexer, name_resolution, parser, type_resolution};
use insta::{assert_snapshot, glob};
use std::{fs, path::Path};

fn test_util(path: &Path, spans: bool, variables: bool, types: bool) -> String {
    let code = fs::read_to_string(path).unwrap();
    let lexer = lexer::lex(&code);
    let mut ast = parser::parse(lexer);

    let variables = if variables | types {
        Some(name_resolution::visit(&mut ast))
    } else {
        None
    };

    let types = if types {
        Some(type_resolution::visit(
            &mut ast,
            variables.as_ref().unwrap(),
        ))
    } else {
        None
    };

    // let _module = spirv_generation::visit(&mut ast, &vars, &type_table);

    debug_print::visit(&ast, true, spans, variables.as_ref(), types.as_ref())
}

#[test]
fn parser() {
    glob!(r"insta_parser\*.coff", |path| {
        assert_snapshot!(test_util(path, false, false, false));
    });
}

#[test]
fn parser_spans() {
    glob!(r"insta_parser_spans\*.coff", |path| {
        assert_snapshot!(test_util(path, true, false, false));
    });
}

#[test]
fn variables() {
    glob!(r"insta_variables\*.coff", |path| {
        assert_snapshot!(test_util(path, false, true, false));
    });
}

#[test]
fn types() {
    glob!(r"insta_types\*.coff", |path| {
        assert_snapshot!(test_util(path, false, false, true));
    });
}
