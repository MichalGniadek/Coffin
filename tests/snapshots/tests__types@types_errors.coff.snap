---
source: tests/tests.rs
expression: "debug_print::visit(&ast, true, false, None, Some(&types))"

---
fun a (arg1 : NonExistant, arg2 : int)  {
	($arg2: error = {
		(#2: int + #2: int): int
	}: int): void
	($arg2: error = {
		($arg2: error = #3: int): void
	}: void): void
	(#3: int + {
		
	}: void): error
}: error

===
Undeclared type. [12..23]
Mismatched type. Expected int, got void. [67..77]
Wrong types for '+', int [83..84] and void [87..89].
